### PR TITLE
support software simulated target for testing  and debugging

### PR DESCRIPTION
**Pull request by rizlik**
This PR introduces a new target `sim`.  This creates an executable ELF wolfboot image `wolfboot.elf`. Internal and optionally and external flash are simulated and backed by regular files. This can be useful for testing and debugging without an actual hardware board.

Example:
```
cp ./config/examples/sim.config .config
make

# create the file internal_flash.dd with firmware v1 on the boot partition and
# firmware v2 on the update partition
make test-sim-internal-flash-with-update
# it should print 1
./wolfboot.elf success get_version
# trigger an update
./wolfboot.elf update_trigger
# it should print 2
./wolfboot.elf success get_version
# it should print 2
./wolfboot.elf success get_version
```

### CODE REVIEW

File: hal/sim.c [hunk 1]
```diff
@@ -0,0 +1,194 @@
+/* dummy.c
```
Comment (danielinux) [inline hal/sim.c:1]: update filename in comment

File: test-app/app_sim.c [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/* main.c
```
Comment (danielinux) [inline test-app/app_sim.c:1]: update filename in comment

File: test-app/app_sim.c [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * Test bare-metal boot-led-on application
+ *
+ * Copyright (C) 2021 wolfSSL Inc.
```
Comment (danielinux) [inline test-app/app_sim.c:5]: (c) 2022

File: hal/sim.c [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+ *
+ * Copyright (C) 2021 wolfSSL Inc.
```
Comment (danielinux) [inline hal/sim.c:3]: (c) 2022

**Review by danielinux — CHANGES_REQUESTED** [review]
Very nice feature, allows for very quick tests.

Check some minor notes on comments.

Please consider adding `*.dd` to `.gitignore`

**Comment by rizlik** [comment]
comment addressed
fixed a small bug in `test.mk`

**Review by danielinux — APPROVED** [review]

**Pull request merged**